### ISSUE TITLE
mm_grounding_dino num_class=256?

### ISSUE THREAD

**Issue by sipie800** [open-mmlab/mmdetection#11765]
configs\mm_grounding_dino\grounding_dino_swin-t_pretrain_obj365.py,

in this cfg, the num_classes in dino head is 256, and then this is inherited by all cfgs.

why is 256 for 365 classes in object365? How does it deal with the classes?

**Comment by AwePhD** [comment]
256 because this is the (max) number of tokens in the text. GroundingDINO follows GLIP reformulation of the detection problem. Previously the number of classes were restricted. Now, the model computes some features on the text and it should match the "content features" of the DINO model. 
During training, there is a `positive_map` for each sample to decide which part of the text represents which region. Note the equivalence, before a region was represented by a single category, now by a group of features. By the way, BLIP2 (if I remember correctly) experienced that these two formulations lead to the same performance. This way, the model should maximize the scores, similarity = scalar product, between the positive text features and their content features.
GroundingDINO is an application of BLIP papers on the DETR detectors, DINO in this case.

**Comment by sipie800** [comment]
> 256 because this is the (max) number of tokens in the text. GroundingDINO follows GLIP reformulation of the detection problem. Previously the number of classes were restricted. Now, the model computes some features on the text and it should match the "content features" of the DINO model. During training, there is a `positive_map` for each sample to decide which part of the text represents which region. Note the equivalence, before a region was represented by a single category, now by a group of features. By the way, BLIP2 (if I remember correctly) experienced that these two formulations lead to the same performance. This way, the model should maximize the scores, similarity = scalar product, between the positive text features and their content features. GroundingDINO is an application of BLIP papers on the DETR detectors, DINO in this case.

Thanks for the detailed explaination. So the num_classes here does not act as the vallina num_class of a supervised classification head which is trained by CE loss, right?

**Comment by AwePhD** [comment]
Kind of yes and not. 
To not be confusing, it is the **same** because the model is trained with FocalLoss (= better CELoss). It has exactly the same role. 
It is **different** because these num_classes are not static classes.

Vanilla detection classification is done by some additional layers, a classification head. The final features are softmaxed and we can compute a cross entropy. If there are 256 classes, there will be 256 scores. score 1 is for class 1, score 2 is for class 2 etc.. The `target` of these scores would be one of the class, note `i` between 1 and 256. CELoss will penalize any high scores in a different class than `i`.

Visual Grounded classification is the same, expect that the classes are not static, classic detection is being called *closed*. A text is a caption about objects in the image. The text has up to 256 tokens, because BERT Base has a context length of 256. For instance, the object 1 is represented by the tokens 1, 2, 3 in the text of the image. 
The prediction of the model are "content features" and "spatial information". "spatial information" is another story, we do not care for classification. The content features are the **same**[1] thing as the output of the classification head of vanilla detector. Then, we want to produce the score for each "class" right? We just do a scalar product[2]  between our content features, which have 256 dimension features, with the 256[3] tokens, each of them have 256 dimensions. Basically this is a matrix product between `content_queries.shape` -> (256) `tokens_encoded`[4] (256, 256). 
It gives `scores.shape` (256). You can softmax (or sigmoid, I do not remember) the result. It gives the scores of the content queries over the tokens. Your FocalLoss (CELoss would be the same) is used in a binary classification case: there are positive tokens and negative tokens; the score should be high in the former case and low in the latter. 

With this and reading the reformulation part of the GLIP paper, it should be fine.

[1]: It's not. Classic detectors, such as DINO, have a prediction head, so actual learnable parameters to output the final features. GDINO takes the raw output of the decoder, without projection.
[2]: You might add a normalization, see grounding_dino_head.py:ContrastiveEmbed.
[3]: This might be confusing, there are 256 tokens with a features dimension equals 256 features, each. 
[4]: `memory_text` in the code. They are "encoded" by the BiAttention of the Encoder which follows BERT.

**Comment by sipie800** [comment]
> Kind of yes and not. To not be confusing, it is the **same** because the model is trained with FocalLoss (= better CELoss). It has exactly the same role. It is **different** because these num_classes are not static classes.
> 
> Vanilla detection classification is done by some additional layers, a classification head. The final features are softmaxed and we can compute a cross entropy. If there are 256 classes, there will be 256 scores. score 1 is for class 1, score 2 is for class 2 etc.. The `target` of these scores would be one of the class, note `i` between 1 and 256. CELoss will penalize any high scores in a different class than `i`.
> 
> Visual Grounded classification is the same, expect that the classes are not static, classic detection is being called _closed_. A text is a caption about objects in the image. The text has up to 256 tokens, because BERT Base has a context length of 256. For instance, the object 1 is represented by the tokens 1, 2, 3 in the text of the image. The prediction of the model are "content features" and "spatial information". "spatial information" is another story, we do not care for classification. The content features are the **same**[1] thing as the output of the classification head of vanilla detector. Then, we want to produce the score for each "class" right? We just do a scalar product[2] between our content features, which have 256 dimension features, with the 256[3] tokens, each of them have 256 dimensions. Basically this is a matrix product between `content_queries.shape` -> (256) `tokens_encoded`[4] (256, 256). It gives `scores.shape` (256). You can softmax (or sigmoid, I do not remember) the result. It gives the scores of the content queries over the tokens. Your FocalLoss (CELoss would be the same) is used in a binary classification case: there are positive tokens and negative tokens; the score should be high in the former case and low in the latter.
> 
> With this and reading the reformulation part of the GLIP paper, it should be fine.
> 
> [1]: It's not. Classic detectors, such as DINO, have a prediction head, so actual learnable parameters to output the final features. GDINO takes the raw output of the decoder, without projection. [2]: You might add a normalization, see grounding_dino_head.py:ContrastiveEmbed. [3]: This might be confusing, there are 256 tokens with a features dimension equals 256 features, each. [4]: `memory_text` in the code. They are "encoded" by the BiAttention of the Encoder which follows BERT.

I see. Kind of the grounding classification is projecting the vision feature to a "token space" supported by token positions and also attended by image-text matching.
I'm new in grounding detection area. Is there any work that matching the sentence embedding itself to a ROI-like thing? IMO the semantic in a grounding caption is not really representing the semantic in a image, at least not always is. There is an intuition that detection task is somehow a local cognition rather than a global analyzing though global information does help. 
In many scenario a human don't care if a man is by the side of a horse. If we do care, it might be better to locate them and analyze the distance (which is yet not done by grounding and can be hard if trying). Shall we make locating and semantation more decoupled ?

**Comment by AwePhD** [comment]
I missed the notification, sorry for the delay.

> Kind of the grounding classification is **projecting the vision feature to a "token space"** supported by token positions and also attended by image-text matching.

Just to be precise, there are two main spaces that the model learns: the `memory_text` and the `content_queries`. Both have text and image information. The latter will have positional information compared to the former. The contrastive loss aims to *align* these spaces, it's not a projection per se. 

> Is there any work that matching the sentence embedding itself to a ROI-like thing? 

I am not sure to understand. GroundingDINO and Visual Grounding models (GLIP, DetCLIP, ...) perform this. In GroundingDINO there is an attention interaction between the query (content + position) and the encoder output. For the `loss` computation, there are separate classification (image / text matching) and localization losses. For details see Figure 5 of DAB-DETR, mainly.

> There is an intuition that detection task is somehow a local cognition rather than a global analyzing though global information does help.

This intuition could be what motivated the researchers to use a *sub-sentence* representation. Basically, the self-attention in the text encoder is only looking to text related to one region in an image. See section 3.4 of GroundingDINO. The paper claims that it removes the interaction between unrelated part of the caption, I disagree with this claim. Since other attentions (bi attention and decoder attention), the token i can have access to some information probed in context j from another layer. But, this is just a detail[1], I agree that this sub-sentence representation helps to isolate the region of the image with narrower part of the caption.

> Shall we make locating and semantation more decoupled?

Open question. Some experiments on DetCLIP and GroundingDINO architectures can be interesting. The direction of the """big""" papers seem to be about how to feed more data with self-supervised / generative model / exhaustive pipelines[2]. The thing that I like the most with DETR-like models is their palpable intuitions. GroundingDINO, DB-DETR, Deformable DETR and DAB-DETR arguably introduce new features from simple intuitions. Therefore, GroundingDINO might benefit from a study of this coupling (how to measure it, what parts of the network influence it) and how to encourage / limit it. It would be an interesting direction to explore. I am not 100% confident it would lead to something though.

[1]: The context of the sub sentence is when the caption is a concatenation of multiple categories or a REC. There is no reason that the model needs to encourage the leak of information. For instance, the caption "A man is next to a horse" and the sub-sentences would be "A man" and "a horse". To learn to locate and "to understand" the sentence, it seems OK to use a mask in the self-attention.

[2]: I think this is an interesting approach, it's very VL**P** oriented.

**Comment by xiexie123** [comment]
> Kind of yes and not. To not be confusing, it is the **same** because the model is trained with FocalLoss (= better CELoss). It has exactly the same role. It is **different** because these num_classes are not static classes.
> 
> Vanilla detection classification is done by some additional layers, a classification head. The final features are softmaxed and we can compute a cross entropy. If there are 256 classes, there will be 256 scores. score 1 is for class 1, score 2 is for class 2 etc.. The `target` of these scores would be one of the class, note `i` between 1 and 256. CELoss will penalize any high scores in a different class than `i`.
> 
> Visual Grounded classification is the same, expect that the classes are not static, classic detection is being called _closed_. A text is a caption about objects in the image. The text has up to 256 tokens, because BERT Base has a context length of 256. For instance, the object 1 is represented by the tokens 1, 2, 3 in the text of the image. The prediction of the model are "content features" and "spatial information". "spatial information" is another story, we do not care for classification. The content features are the **same**[1] thing as the output of the classification head of vanilla detector. Then, we want to produce the score for each "class" right? We just do a scalar product[2] between our content features, which have 256 dimension features, with the 256[3] tokens, each of them have 256 dimensions. Basically this is a matrix product between `content_queries.shape` -> (256) `tokens_encoded`[4] (256, 256). It gives `scores.shape` (256). You can softmax (or sigmoid, I do not remember) the result. It gives the scores of the content queries over the tokens. Your FocalLoss (CELoss would be the same) is used in a binary classification case: there are positive tokens and negative tokens; the score should be high in the former case and low in the latter.
> 
> With this and reading the reformulation part of the GLIP paper, it should be fine.
> 
> [1]: It's not. Classic detectors, such as DINO, have a prediction head, so actual learnable parameters to output the final features. GDINO takes the raw output of the decoder, without projection. [2]: You might add a normalization, see grounding_dino_head.py:ContrastiveEmbed. [3]: This might be confusing, there are 256 tokens with a features dimension equals 256 features, each. [4]: `memory_text` in the code. They are "encoded" by the BiAttention of the Encoder which follows BERT.

good answer. 感谢精彩回复。我还有一个问题，groundingdino基于dino，使用了denoising training.  即代码里使用了CdnQueryGenerator， CdnQueryGenerator的构造是基于class num的。所以，不管是pretrain 还是finetune groundingdino, CdnQueryGenerator中num_classes都应该是256 对吧，无论实际dataset的类别是多少。我理解对吗？

**Comment by AwePhD** [comment]
I do not speak Chinese, here an automatic translation, so you know what I understand from your answer:

> good answer. Thanks for the wonderful reply. I have another question. Groundingdino is based on dino and uses denoising training. That is, CdnQueryGenerator is used in the code. The construction of CdnQueryGenerator is based on class num. So, whether it is pretraining or finetuning groundingdino, num_classes in CdnQueryGenerator should be 256, right, no matter how many categories the actual dataset has. Am I right?

Yes you are right. CDN in DINO added the negative queries compared to DN-DETR. Thus, the negative should have a category "unknown". In the context of DINO on the COCO dataset, it means to add a special category "unknown" on top of the existing ones.
In Grounding DINO, it learns one embedding, one token, for negative CDN query. So 256+1 embeddings, 256 for positive and + 1 for negative. It's similar to learn a specific token[1] `[UNK]` for negative CDN query.

There might be some questions raised by the CDN in Grounding DINO. The main question, for me, is the embedding versus the text caption. Embedding are learned and static, there are `num_classes=256` of them in Grounding DINO. The first tokens might be "a dog" or "a tree" or anything else based on the frame/text. How the model can benefit from such a "restricted" embedding? I think there might be a room for improvement. Maybe it has been done already, maybe I misunderstood something, maybe some people tried it and it did not lead to anything. I would be interested to read on that. 
Also, one embedding for all negative is even more restricted. When regions of an image become even more semantic rich with multiple tokens, how a static learned embedding can be a good training?
The questions do not have a straight answer, since the denoising training should be easy and stable. If too much complexity is added, then it would hurt the training's convergence.

[1]: My sentence is a bit confusing. "a specific token" refers to a query embedding which is used as a token. In my answer I use token and embedding to express the same thing, since in the case of `CdnQueryGenerator` this is the same thing. Token is more meaningful for understanding the subsequent transformations of the embedding in the decoder and the role of these outputs in the loss.

**Comment by xiexie123** [comment]
> I do not speak Chinese, here an automatic translation, so you know what I understand from your answer:
> 
> > good answer. Thanks for the wonderful reply. I have another question. Groundingdino is based on dino and uses denoising training. That is, CdnQueryGenerator is used in the code. The construction of CdnQueryGenerator is based on class num. So, whether it is pretraining or finetuning groundingdino, num_classes in CdnQueryGenerator should be 256, right, no matter how many categories the actual dataset has. Am I right?
> 
> Yes you are right. CDN in DINO added the negative queries compared to DN-DETR. Thus, the negative should have a category "unknown". In the context of DINO on the COCO dataset, it means to add a special category "unknown" on top of the existing ones. In Grounding DINO, it learns one embedding, one token, for negative CDN query. So 256+1 embeddings, 256 for positive and + 1 for negative. It's similar to learn a specific token[1] `[UNK]` for negative CDN query.
> 
> There might be some questions raised by the CDN in Grounding DINO. The main question, for me, is the embedding versus the text caption. Embedding are learned and static, there are `num_classes=256` of them in Grounding DINO. The first tokens might be "a dog" or "a tree" or anything else based on the frame/text. How the model can benefit from such a "restricted" embedding? I think there might be a room for improvement. Maybe it has been done already, maybe I misunderstood something, maybe some people tried it and it did not lead to anything. I would be interested to read on that. Also, one embedding for all negative is even more restricted. When regions of an image become even more semantic rich with multiple tokens, how a static learned embedding can be a good training? The questions do not have a straight answer, since the denoising training should be easy and stable. If too much complexity is added, then it would hurt the training's convergence.
> 
> [1]: My sentence is a bit confusing. "a specific token" refers to a query embedding which is used as a token. In my answer I use token and embedding to express the same thing, since in the case of `CdnQueryGenerator` this is the same thing. Token is more meaningful for understanding the subsequent transformations of the embedding in the decoder and the role of these outputs in the loss.

Thanks for your answer. I think I get your point and your questions is also mine. I think dn_loss_cls of CDN  in  Grounding DINO does not make sense.  So in my training, I only use dn_loss_bbox and dn_loss_iou.

**Comment by AwePhD** [comment]
Thank you for the feedback. In my experiment, `dn_loss_cls` is converging well. I observed, by experiments on a tiny/debug dataset, that removing any CDN loss hurts the convergence. I should mention that I use GDINO for a different (harder) task than localization+classification. I did not try a training on a "normal" size dataset. So this is interesting to know that the training can be done without `dn_loss_cls`.
